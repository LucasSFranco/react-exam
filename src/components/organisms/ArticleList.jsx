import { useCallback, useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import ArticleListItem from '../molecules/ArticleListItem'
import ArticleListItemSkeleton from '../molecules/ArticleListItemSkeleton'
import ArticleModal from '../molecules/ArticleModal'
import {
  getAllArticles,
  getAllArticlesCount,
  getTopicArticles,
  getTopicArticlesCount,
  selectActiveTab,
  selectArticles,
  selectCount,
  selectSynching,
} from '../../store/slices/newsSlice'

import styled from 'styled-components'

function ArticleList() {
  const dispatch = useDispatch()
  const activeTab = useSelector(selectActiveTab)
  const articles = useSelector(selectArticles)
  const count = useSelector(selectCount)
  const synching = useSelector(selectSynching)

  const loader = useRef(
    [...new Array(10).keys()]
      .map(key => <ArticleListItemSkeleton key={key} />)
  )

  useEffect(() => {
    if(!synching && activeTab && !articles[activeTab]) {
      if(activeTab !== 'all news') {
        dispatch(getTopicArticles(activeTab))
        dispatch(getTopicArticlesCount(activeTab))
      } else {
        dispatch(getAllArticles())
        dispatch(getAllArticlesCount())
      }
    }
  }, [activeTab, articles, dispatch, synching])

  const observer = useRef()
  const lastArticleRef = useCallback(node => {
    observer.current = new IntersectionObserver(entries => {
      const [entry] = entries

      if(entry.isIntersecting && count[activeTab] !== articles[activeTab].length) {
        if(activeTab !== 'all news')
          dispatch(getTopicArticles(activeTab))
        else
          dispatch(getAllArticles())

        observer.current.disconnect()
      }
    })

    if(node) observer.current.observe(node)

  }, [activeTab, articles, count, dispatch])

  return (
    <Self className="article list">
      {
        !synching && (articles[activeTab] || []).map(
          (article, pos) => (
            <ArticleModal
              key={article.id}
              article={article}
            >
              {
                pos + 1 === articles[activeTab].length ? (
                  <ArticleListItem
                    forwardRef={lastArticleRef}
                    article={article}
                  />
                ) : (
                  <ArticleListItem
                    article={article}
                  />
                )
              }
            </ArticleModal>
          )
        )
      }

      {
        count[activeTab] !== (articles[activeTab]?.length || []) && loader.current
      }

    </Self>
  )
}

const Self = styled.section`
  border-top: 2px solid var(--accent-primary);
  padding: 3rem 0;
`

export default ArticleList
