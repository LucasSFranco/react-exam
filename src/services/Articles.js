import api from '@/services/_api'
import db from '@/services/_dexie'

import Article from '@/models/Article'

class Articles {

  static topics = ['science', 'technology']

  static async synchronize() {
    await Promise.all(
      Articles.topics.map(
        async topic => {
          const { data } = await api.get(`${topic}.json`)

          const articles = data.results
            .map(result => new Article({ topic, ...result }))

          await Promise.all(
            articles.map(async article => {
              const alreadyExists = await db.articles.get({ uri: article.uri })

              if(alreadyExists) return

              await db.articles.put(article)
            })
          )
        }
      )
    )

    console.log('Synchronized!')
  }

  static async getAll(offset = 0) {
    const allArticles = await db.articles
      .reverse()
      .offset(offset)
      .limit(10)
      .toArray()

    // await new Promise((resolve) => setTimeout(() => {resolve()}, 1000))

    return allArticles
  }

  static async getBy(topic, offset = 0) {
    const topicArticles = await db.articles
      .where('topic')
      .equals(topic)
      .reverse()
      .offset(offset)
      .limit(10)
      .toArray()

    // await new Promise((resolve) => setTimeout(() => {resolve()}, 1000))

    return topicArticles
  }

  static async getAllCount() {
    return db.articles
      .count()
  }

  static async getTopicCount(topic) {
    return db.articles
      .where('topic')
      .equals(topic)
      .count()
  }

}

export default Articles
