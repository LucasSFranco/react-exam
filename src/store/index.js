import { configureStore } from '@reduxjs/toolkit'

import newsReducer from '@/store/slices/newsSlice'

export default configureStore({
  reducer: { news: newsReducer },
})
