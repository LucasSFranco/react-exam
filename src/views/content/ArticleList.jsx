import { useCallback, useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useParams } from 'react-router'

import ArticleListItemSkeleton from '@/views/content/ArticleListItemSkeleton'
import ArticleListItem from '@/views/content/ArticleListItem'
import ArticleModal from '@/views/content/ArticleModal'
import {
  getAllArticles,
  getAllArticlesCount,
  getArticlesByTopic,
  getTopicArticlesCount,
  selectActiveTab,
  selectArticles,
  selectCount,
  selectSynching,
} from '@/store/slices/appSlice'

import styled from 'styled-components'

function ArticleList() {
  const dispatch = useDispatch()
  const activeTab = useSelector(selectActiveTab)
  const articles = useSelector(selectArticles)
  const count = useSelector(selectCount)
  const synching = useSelector(selectSynching)

  useEffect(() => {
    if(!synching) {
      if(activeTab !== 'all news') {
        dispatch(getArticlesByTopic(activeTab))
        dispatch(getTopicArticlesCount(activeTab))
      } else {
        dispatch(getAllArticles())
        dispatch(getAllArticlesCount())
      }
    }
  }, [dispatch, synching, activeTab])

  const observer = useRef()
  const lastArticleRef = useCallback(node => {

    observer.current = new IntersectionObserver(entries => {
      const [entry] = entries

      if(entry.isIntersecting && count[activeTab] !== articles[activeTab].length) {
        if(activeTab !== 'all news')
          dispatch(getArticlesByTopic(activeTab))
        else
          dispatch(getAllArticles())

        observer.current.disconnect()
      }
    })

    if(node) observer.current.observe(node)
  }, [dispatch, articles, count, activeTab])

  return (
    <Self>
      {
        !synching && articles?.[activeTab] && articles[activeTab].map(
          (article, id) => (
            <ArticleModal
              key={article.id}
              article={article}
            >
              {
                id + 1 === articles[activeTab].length ? (
                  <ArticleListItem
                    forwardRef={lastArticleRef}
                    article={article}
                  />
                ) : (
                  <ArticleListItem
                    article={article}
                  />
                )
              }
            </ArticleModal>
          )
        )
      }

      {
        (synching || count[activeTab] !== articles?.[activeTab]?.length) && [...new Array(10).keys()].map((arg, id) =>
          <ArticleListItemSkeleton key={id} />
        )
      }

    </Self>
  )
}

const Self = styled.section`
  border-top: 2px solid var(--accent-primary);
  padding: 3rem 0;
`

export default ArticleList
