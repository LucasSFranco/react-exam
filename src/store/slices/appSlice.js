import { createSlice, createAsyncThunk } from '@reduxjs/toolkit'

import Articles from '@/services/Articles'

export const getAllArticlesCount = createAsyncThunk('app/getAllArticlesCount',
  async () => {
    const allArticlesCount = await Articles.getAllCount()

    return { 'all news': allArticlesCount }
  }
)

export const getTopicArticlesCount = createAsyncThunk('app/getTopicArticlesCount',
  async (topic) => {
    const topicArticlesCount = await Articles.getTopicCount(topic)

    return { [topic]: topicArticlesCount }
  }
)

export const getAllArticles = createAsyncThunk('app/getAllArticles',
  async (args, { getState }) => {
    const { articles } = getState().app

    const allArticles = await Articles.getAll(articles?.['all news']?.length || 0)

    return { 'all news': [...(articles?.['all news'] || []), ...allArticles] }
  }
)

export const getArticlesByTopic = createAsyncThunk('app/getArticlesByTopic',
  async (topic, { getState }) => {
    const { articles } = getState().app

    const topicArticles = await Articles.getBy(topic, articles?.[topic]?.length || 0)

    return { [topic]: [...(articles?.[topic] || []), ...topicArticles] }
  }
)

export const appSlice = createSlice({
  name: 'app',
  initialState: {
    tabs: ['all news', ...Articles.topics],
    activeTab: 'all news',
    articles: {},
    count: {},
  },
  reducers: {
    changeActiveTab: (state, action) => {
      state.activeTab = action.payload
    },
  },
  extraReducers: {
    [getAllArticles.fulfilled]: (state, action) => {
      state.articles = { ...state.articles, ...action.payload }
    },
    [getArticlesByTopic.fulfilled]: (state, action) => {
      state.articles = { ...state.articles, ...action.payload }
    },
    [getAllArticlesCount.fulfilled]: (state, action) => {
      state.count = { ...state.count, ...action.payload }
    },
    [getTopicArticlesCount.fulfilled]: (state, action) => {
      state.count = { ...state.count, ...action.payload }
    },
  }
})

export const { changeActiveTab } = appSlice.actions

export const selectCount = state => state.app.count
export const selectTabs = state => state.app.tabs
export const selectActiveTab = state => state.app.activeTab
export const selectArticles = state => state.app.articles

export default appSlice.reducer
