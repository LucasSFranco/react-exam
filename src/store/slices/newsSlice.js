import { createAsyncThunk, createSlice } from '@reduxjs/toolkit'

import Articles from '../../services/Articles'

export const getAllArticles = createAsyncThunk('news/getAllArticles',
  async (arg, { getState }) => {
    const { articles } = getState().news

    const allArticles = await Articles.getAll(articles['all news']?.length || 0)

    return { 'all news': [...(articles['all news'] || []), ...allArticles] }
  }
)

export const getAllArticlesCount = createAsyncThunk('news/getAllArticlesCount',
  async () => {
    const allArticlesCount = await Articles.getAllCount()

    return { 'all news': allArticlesCount }
  }
)

export const getTopicArticles = createAsyncThunk('news/getTopicArticles',
  async (topic, { getState }) => {
    const { articles } = getState().news

    const topicArticles = await Articles.getTopic(topic, articles[topic]?.length || 0)

    return { [topic]: [...(articles[topic] || []), ...topicArticles] }
  }
)

export const getTopicArticlesCount = createAsyncThunk('news/getTopicArticlesCount',
  async (topic) => {
    const topicArticlesCount = await Articles.getTopicCount(topic)

    return { [topic]: topicArticlesCount }
  }
)

export const sync = createAsyncThunk('news/sync',
  async (arg, { rejectWithValue }) => {
    try {
      await Articles.sync()
    } catch(err) {
      rejectWithValue(err)
    }
  }
)

export const newsSlice = createSlice({
  name: 'news',
  initialState: {
    activeTab: null,
    articles: {},
    count: {},
    synching: false,
    tabs: ['all news', ...Articles.topics],
  },
  reducers: {
    changeActiveTab: (state, action) => {
      state.activeTab = action.payload
    },
  },
  extraReducers: {
    [getAllArticles.fulfilled]: (state, action) => {
      state.articles = { ...state.articles, ...action.payload }
    },
    [getAllArticlesCount.fulfilled]: (state, action) => {
      state.count = { ...state.count, ...action.payload }
    },
    [getTopicArticles.fulfilled]: (state, action) => {
      state.articles = { ...state.articles, ...action.payload }
    },
    [getTopicArticlesCount.fulfilled]: (state, action) => {
      state.count = { ...state.count, ...action.payload }
    },
    [sync.pending]: (state) => {
      state.synching = true
    },
    [sync.fulfilled]: (state) => {
      state.synching = false
    },
    [sync.reject]: (state) => {
      state.synching = false
    },
  },
})

export const { changeActiveTab } = newsSlice.actions

export const selectActiveTab = state => state.news.activeTab
export const selectArticles = state => state.news.articles
export const selectCount = state => state.news.count
export const selectSynching = state => state.news.synching
export const selectTabs = state => state.news.tabs

export default newsSlice.reducer
