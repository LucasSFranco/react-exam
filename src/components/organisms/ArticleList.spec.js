import React from 'react'
import { Provider } from 'react-redux'
import { render, screen } from '@testing-library/react'

import ArticleList from './ArticleList'
import Articles from '../../services/Articles'
import store from '../../store'
import { changeActiveTab } from "../../store/slices/newsSlice"
import articles from '../../../tests/fixtures/articles.json'

jest.mock('../../services/Articles')

describe('Article List', () => {

  beforeEach(async () => {
    const mockIntersectionObserver = jest.fn()

    mockIntersectionObserver.mockReturnValue({
      observe: () => null,
      disconnect: () => null,
    })

    window.IntersectionObserver = mockIntersectionObserver
  })

  it('properly renders all articles', async () => {
    Articles.getAll.mockResolvedValue(articles)
    Articles.getAllCount.mockResolvedValue(articles.length)

    const { container } = render(
      <Provider store={store}>
        <ArticleList />
      </Provider>
    )

    expect(await screen.findAllByTestId('article-list-item'))
      .toHaveLength(articles.length)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('properly renders topic articles', async () => {
    const technologyArticles = articles
      .filter(article => article.topic === 'technology')

    Articles.getTopic.mockResolvedValue(technologyArticles)
    Articles.getTopicCount.mockResolvedValue(technologyArticles.length)

    store.dispatch(changeActiveTab('technology'))

    const { container } = render(
      <Provider store={store}>
        <ArticleList />
      </Provider>
    )

    expect(await screen.findAllByTestId('article-list-item'))
      .toHaveLength(technologyArticles.length)

    expect(container.firstChild).toMatchSnapshot()
  })

  // it('properly renders topic articles', async () => {
  //   const technologyArticles = articles
  //     .filter(article => article.topic === 'technology')

  //   Articles.getTopic.mockResolvedValue(technologyArticles)
  //   Articles.getTopicCount.mockResolvedValue(technologyArticles.length)

  //   store.dispatch(changeActiveTab('technology'))

  //   const { container } = render(
  //     <Provider store={store}>
  //       <ArticleList />
  //     </Provider>
  //   )

  //   expect(await screen.findAllByTestId('article-list-item'))
  //     .toHaveLength(technologyArticles.length)

  //   expect(container.firstChild).toMatchSnapshot()
  // })
})
