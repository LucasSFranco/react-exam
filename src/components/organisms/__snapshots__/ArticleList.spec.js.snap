// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article List properly renders all articles 1`] = `
<section
  class="sc-jSFjdj bOIGTT article list"
>
  <div
    class="sc-bdnxRM leLQZM article list__item"
    data-testid="article-list-item"
  >
    <div
      class="accessory"
    >
      <time>
        May 19, 2021
      </time>
    </div>
    <div
      class="main"
    >
      <figure>
        <picture
          style="padding-top: 66.650390625%;"
        >
          <source
            media="(min-width: 768px)"
            srcset="https://static01.nyt.com/images/2021/05/19/world/19china-mars01/merlin_187780170_e6f955da-f88e-4dfa-9513-845ef968ceda-mediumThreeByTwo210.jpg"
          />
          <img
            alt="Coverage of the Mars landing on a screen in Beijing on Saturday. The first images from the lander were released on Wednesday."
            src="https://static01.nyt.com/images/2021/05/19/world/19china-mars01/merlin_187780170_e6f955da-f88e-4dfa-9513-845ef968ceda-superJumbo.jpg"
          />
        </picture>
        <figcaption>
          Mark Schiefelbein/Associated Press
        </figcaption>
      </figure>
      <h6 />
      <h3>
        China Releases First Photos From Its Mars Lander
      </h3>
      <p>
        The country’s space agency said that the craft’s components, and those of the rover it would soon release, had “deployed in place normally.”
      </p>
      <span>
        By Steven Lee Myers
      </span>
    </div>
  </div>
  <div
    class="sc-bdnxRM leLQZM article list__item"
    data-testid="article-list-item"
  >
    <div
      class="accessory"
    >
      <time>
        May 19, 2021
      </time>
    </div>
    <div
      class="main"
    >
      <figure>
        <picture
          style="padding-top: 66.650390625%;"
        >
          <source
            media="(min-width: 768px)"
            srcset="https://static01.nyt.com/images/2021/05/19/business/19whatsapp-israel1/merlin_187693269_c6ae324b-0fb1-4a51-bcb0-311f0faafb89-mediumThreeByTwo210.jpg"
          />
          <img
            alt="An Israeli man at a restaurant that was attacked last week in the city of Bat Yam."
            src="https://static01.nyt.com/images/2021/05/19/business/19whatsapp-israel1/merlin_187693269_c6ae324b-0fb1-4a51-bcb0-311f0faafb89-superJumbo.jpg"
          />
        </picture>
        <figcaption>
          Gil Cohen-Magen/Agence France-Presse — Getty Images
        </figcaption>
      </figure>
      <h6 />
      <h3>
        Mob Violence Against Palestinians in Israel Is Fueled by Groups on WhatsApp
      </h3>
      <p>
        Jewish extremists have formed more than 100 new groups on the Facebook-owned encrypted messaging app in recent days to target attacks.
      </p>
      <span>
        By Sheera Frenkel
      </span>
    </div>
  </div>
  <div
    class="sc-bdnxRM leLQZM article list__item"
    data-testid="article-list-item"
  >
    <div
      class="accessory"
    >
      <time>
        May 18, 2021
      </time>
    </div>
    <div
      class="main"
    >
      <figure>
        <picture
          style="padding-top: 69.384765625%;"
        >
          <source
            media="(min-width: 768px)"
            srcset="https://static01.nyt.com/images/2021/05/19/business/19att-print/19att-print-mediumThreeByTwo210.jpg"
          />
          <img
            alt="Alexander Graham Bell, with the first telephone, circa 1876."
            src="https://static01.nyt.com/images/2021/05/19/business/19att-print/19att-print-superJumbo.jpg"
          />
        </picture>
        <figcaption>
          Ullstein Bild, via Getty Images
        </figcaption>
      </figure>
      <h6 />
      <h3>
        After Media Detour, AT&T Confronts Old Problems
      </h3>
      <p>
        Following its spinoff of WarnerMedia, AT&T is back to primarily being a wireless company, with all of the challenges that entails.
      </p>
      <span>
        By Edmund Lee and Lauren Hirsch
      </span>
    </div>
  </div>
  <div
    class="sc-bdnxRM leLQZM article list__item"
    data-testid="article-list-item"
  >
    <div
      class="accessory"
    >
      <time>
        May 18, 2021
      </time>
    </div>
    <div
      class="main"
    >
      <figure>
        <picture
          style="padding-top: 66.650390625%;"
        >
          <source
            media="(min-width: 768px)"
            srcset="https://static01.nyt.com/images/2021/05/18/us/politics/18dc-china-bill-1/merlin_185149560_9f9e2f5e-7b01-4a88-a4be-67dc94324196-mediumThreeByTwo210.jpg"
          />
          <img
            alt="A semiconductor factory in Nantong, China, in March. The country is the host of a critical supply chains including semiconductors."
            src="https://static01.nyt.com/images/2021/05/18/us/politics/18dc-china-bill-1/merlin_185149560_9f9e2f5e-7b01-4a88-a4be-67dc94324196-superJumbo.jpg"
          />
        </picture>
        <figcaption>
          Agence France-Presse — Getty Images
        </figcaption>
      </figure>
      <h6 />
      <h3>
        Senate Weighs Investing $120 Billion in Science to Counter China
      </h3>
      <p>
        The legislation has drawn bipartisan support amid the coronavirus pandemic as Democrats and Republicans have become increasingly concerned about Beijing’s supply chain dominance.
      </p>
      <span>
        By Catie Edmondson
      </span>
    </div>
  </div>
  <div
    class="sc-bdnxRM leLQZM article list__item"
    data-testid="article-list-item"
  >
    <div
      class="accessory"
    >
      <time>
        May 18, 2021
      </time>
    </div>
    <div
      class="main"
    >
      <figure>
        <picture
          style="padding-top: 66.552734375%;"
        >
          <source
            media="(min-width: 768px)"
            srcset="https://static01.nyt.com/images/2021/05/19/climate/18CLI-PLASTIC-print/18CLI-PLASTIC-mediumThreeByTwo210.jpg"
          />
          <img
            alt="Collecting plastic at a landfill on Sulawesi in Indonesia. Half the world&rsquo;s single-use plastic is made by 20 big companies, researchers said."
            src="https://static01.nyt.com/images/2021/05/19/climate/18CLI-PLASTIC-print/18CLI-PLASTIC-superJumbo.jpg"
          />
        </picture>
        <figcaption>
          Basri Marzuki/NurPhoto, via Getty Images
        </figcaption>
      </figure>
      <h6 />
      <h3>
        Here Is Who’s Behind the Global Surge in Single-Use Plastic
      </h3>
      <p>
        A new report shows that a surprisingly small number of big companies and banks are behind the manufacturing and financing of much of the world’s single-use plastic.
      </p>
      <span>
        By Michael Corkery and Somini Sengupta
      </span>
    </div>
  </div>
</section>
`;
